File: solution/1457____Pseudo-Palindromic_Paths_in_a_Binary_Tree.c

int ans = 0;
unsigned mask = 0;

inline bool satisfy() {
        if(!mask)
                return true;
        if(!(mask & (mask - 1)))
                return true;
        return false;
}


// root is guaranteed not null
void helper(struct TreeNode* root) {
        unsigned temp = 1 << root->val;
        mask ^= temp;
        if(root->left)
                helper(root->left);
        if(root->right)
                helper(root->right);
        if(root->left == NULL && root->right == NULL && satisfy()){
                ++ans;
        }
        mask ^= temp;
}

/**
 * @brief 5, 7
 */
int pseudoPalindromicPaths (struct TreeNode* root){
        ans = 0;
        mask = 0;
        if(root == NULL)
                return 0;
        helper(root);
        return ans;
}

File: solution/1457____Pseudo-Palindromic Paths in a Binary Tree.go
package main

import (
	. "model"
)

var ans int = 0
var single_cnt int = 0
var info = [10]bool{false}

func pseudoPalindromicPaths(root *TreeNode) int {
	return _1457(root)
}

func _1457(root *TreeNode) int {
	for i := range info {
		info[i] = false
	}
	if root != nil {
		_1457_recursive_traverse(root)
	}
	return ans
}

func _1457_recursive_traverse(root *TreeNode) {
	info[root.Val] = info[root.Val] != true
	if info[root.Val] {
		single_cnt++
	} else {
		single_cnt--
	}
	if root.Left != nil {
		_1457_recursive_traverse(root.Left)
	}
	if root.Right != nil {
		_1457_recursive_traverse(root.Right)
	}
	if root.Left == nil && root.Right == nil && single_cnt <= 1 { // leaf node, check it
		ans++
	}
	info[root.Val] = info[root.Val] != true
	if info[root.Val] {
		single_cnt++
	} else {
		single_cnt--
	}
}

func main() {

}
